### ISSUE TITLE
services.jupyter: uid/gid are not fixed

### ISSUE THREAD

**Issue by sersorrel** [NixOS/nixpkgs#175674]
### Describe the bug
The jupyter module lets you (reasonably) easily set up a jupyter server, which persists data in /var/lib/jupyter. Unfortunately it doesn't set a static UID or GID for the `jupyter` user or group that it creates by default, which leads to potential difficulties if that user/group are ever allocated a different UID or GID, since the contents of /var/lib/jupyter will now no longer be accessible to the server.

### Steps To Reproduce
1. On NixOS 21.11, set up jupyter like so: 
    ```nix
    {
      services.jupyter.enable = true;
      services.jupyter.password = "''"; # require an auth token printed at startup
      users.users.jupyter.group = "jupyter";
      users.users.jupyter.isSystemUser = true;
    }
    ```
3. Update to NixOS 22.05
4. Now your jupyter doesn't work any more:
    ```
    Jun 01 03:05:53 the jupyter-notebook[1711]: [E 03:05:53.463 NotebookApp] Failed to write server-info to /var/lib/jupyter/.local/share/jupyter/runtime/nbserver-1711.json: [Errno 13] Permission denied: '/var/lib/jupyter/.local/share/jupyter/runtime/nbserver-1711.json'
    Jun 01 03:05:53 the jupyter-notebook[1711]: Traceback (most recent call last):
    Jun 01 03:05:53 the jupyter-notebook[1711]:   File "/nix/store/3dv4p691y9ma7lgj3y4kjbkz42klcx3x-python3.9-notebook-6.4.10/bin/.jupyter-notebook-wrapped", line 9, in <module>
    Jun 01 03:05:53 the jupyter-notebook[1711]:     sys.exit(main())
    Jun 01 03:05:53 the jupyter-notebook[1711]:   File "/nix/store/0vffr5zmkjn9irpwisncg2g0y6j67pzw-python3.9-jupyter_core-4.9.2/lib/python3.9/site-packages/jupyter_core/application.py", line 264, in launch_instance
    Jun 01 03:05:53 the jupyter-notebook[1711]:     return super(JupyterApp, cls).launch_instance(argv=argv, **kwargs)
    Jun 01 03:05:53 the jupyter-notebook[1711]:   File "/nix/store/22lmb6zg2slqpggrg0n00g51jvq4821x-python3.9-traitlets-5.1.1/lib/python3.9/site-packages/traitlets/config/application.py", line 846, in launch_instance
    Jun 01 03:05:53 the jupyter-notebook[1711]:     app.start()
    Jun 01 03:05:53 the jupyter-notebook[1711]:   File "/nix/store/3dv4p691y9ma7lgj3y4kjbkz42klcx3x-python3.9-notebook-6.4.10/lib/python3.9/site-packages/notebook/notebookapp.py", line 2338, in start
    Jun 01 03:05:53 the jupyter-notebook[1711]:     self.write_browser_open_file()
    Jun 01 03:05:53 the jupyter-notebook[1711]:   File "/nix/store/3dv4p691y9ma7lgj3y4kjbkz42klcx3x-python3.9-notebook-6.4.10/lib/python3.9/site-packages/notebook/notebookapp.py", line 2241, in write_browser_open_file
    Jun 01 03:05:53 the jupyter-notebook[1711]:     with open(self.browser_open_file, 'w', encoding='utf-8') as f:
    Jun 01 03:05:53 the jupyter-notebook[1711]: PermissionError: [Errno 13] Permission denied: '/var/lib/jupyter/.local/share/jupyter/runtime/nbserver-1711-open.html'
    Jun 01 03:05:53 the systemd[1]: jupyter.service: Main process exited, code=exited, status=1/FAILURE
    ```
    and all its files are owned by a user and group with no name:
    ```
    $ id jupyter
    uid=993(jupyter) gid=989(jupyter) groups=989(jupyter)
    $ sudo ls -la /var/lib/jupyter/
    [sudo] password for ash:
    total 956
    drwx------  9 jupyter jupyter   4096 May 25 00:16  .
    drwxr-xr-x 26 root    root      4096 Apr 17 03:37  ..
    -rw-r--r--  1     995     992   1434 Apr 17 15:31  2022-04-17-0012-full100-array-2a4935b1ca863a322919f054ff2031bb68b16e83.csv
    -rw-r--r--  1     995     992   1434 Apr 17 15:31  2022-04-17-0013-full100-array-2a4935b1ca863a322919f054ff2031bb68b16e83.csv
    -rw-r--r--  1     995     992   1435 Apr 17 15:31  2022-04-17-0014-full100-array-2a4935b1ca863a322919f054ff2031bb68b16e83.csv
    -rw-r--r--  1     995     992   1435 Apr 17 17:52  2022-04-17-0015-full100-array-2a4935b1ca863a322919f054ff2031bb68b16e83-dirty-without-rewritten-arrays.csv
    -rw-r--r--  1     995     992   1419 Apr 17 17:52  2022-04-17-0016-full100-openmp-f22e1228bc03df284066a23a46f7a72d71e54371-40threads.csv
    -rw-r--r--  1     995     992   1419 Apr 17 17:53  2022-04-17-0017-full100-openmp-f22e1228bc03df284066a23a46f7a72d71e54371-20threads.csv
    -rw-r--r--  1     995     992    397 Apr 19 03:55  2022-04-19-0001-openmp-f22e1228bc03df284066a23a46f7a72d71e54371-allthreadcounts.csv
    -rw-r--r--  1     995     992    424 Apr 25 03:43  2022-04-25-0015-openmp-f22e1228bc03df284066a23a46f7a72d71e54371-1600-like-mpi-threadcounts.csv
    -rw-r--r--  1     995     992    115 Apr 25 17:08  2022-04-25-0016-default-2a4935b1ca863a322919f054ff2031bb68b16e83-oversubscribed.csv
    -rw-r--r--  1     995     992    109 Apr 25 17:10  2022-04-25-0017-default-2a4935b1ca863a322919f054ff2031bb68b16e83-not-oversubscribed.csv
    ...
    ```

### Expected behavior
I shouldn't have to fix the permissions on my jupyter server's files after I upgrade. I guess either it should get allocated a static uid/gid (which is [clearly possible](https://github.com/NixOS/nixpkgs/blob/21443080ad69b4d0b43258abf235bd332ebd4141/nixos/modules/misc/ids.nix)) or it should use [systemd's DynamicUser and StateDirectory](https://0pointer.net/blog/dynamic-users-with-systemd.html).

### Notify maintainers

@aborsu

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.43, NixOS, 22.05 (Quokka), 22.05.342.a634c8f6c1f`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.8.1`
 - channels(ash): `"home-manager-22.05, nixos-unstable-22.11pre381914.f1c167688a6"`
 - channels(root): `"nixos-22.05.342.a634c8f6c1f"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

**Comment by sersorrel** [comment]
I've hacked up the nixpkgs module to support only DynamicUser: https://github.com/sersorrel/config/blob/57c176aa682d98e76ae040018457abd34b99dd4c/hosts/the/jupyter/jupyter.nix

I have no idea how to upstream that in a backwards-compatible way, though.

**Comment by teto** [comment]
> I've hacked up the nixpkgs module to support only DynamicUser: [sersorrel/config@57c176a/hosts/the/jupyter/jupyter.nix](https://github.com/sersorrel/config/blob/57c176aa682d98e76ae040018457abd34b99dd4c/hosts/the/jupyter/jupyter.nix)

How does that work ? could you open a PR so we can discuss the actual changes please ? Looks interesting.

**Comment by sersorrel** [comment]
sorry, I stopped using Jupyter at some point in the past year so wouldn't have the motivation to push a change through. you're welcome to use the code, though; consider the modifications CC0 if that helps.

the interesting things are hardcoding `notebookDir` to `/var/lib/jupyter` and setting `DynamicUser=true` and `StateDirectory="jupyter"`. I think I had to set $HOME for some reason too, though I don't remember why.
